# Exercises I

In [55]:
import numpy as np
import pandas as pd
import matplotlib as plt
import env
from pydataset import data

#### 1. Run python -m pip install pymysql from your terminal to install the mysql client (any folder is fine)
Done

#### 2. cd into your exercises folder for this module and run echo env.py >> .gitignore
Done

#### 3. Create a function named get_db_url. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.



In [56]:
def get_db_url(db_name):
    return (f'mysql+pymysql://{env.user}:{env.password}@{env.host}/{db_name}')
    

#### 4. Use your function to obtain a connection to the employees database.

In [57]:
url = get_db_url('employees')


In [58]:
pd.read_sql('''
    select * from employees
    limit 5;
    ''', get_db_url('employees'))

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

#### 5. Once you have successfully run a query:

####     a. Intentionally make a typo in the database url. What kind of error message do you see?



In [59]:
pd.read_sql('''
    select * from employees
    limit 5;
    ''', get_db_url('employes'))

OperationalError: (pymysql.err.OperationalError) (1044, "Access denied for user 'oneil_2100'@'%' to database 'employes'")
(Background on this error at: https://sqlalche.me/e/14/e3q8)

#### b. Intentionally make an error in your SQL query. What does the error message look like?


In [60]:
pd.read_sql('''
    select * from employes
    limit 5;
    ''', get_db_url('employees'))

ProgrammingError: (pymysql.err.ProgrammingError) (1146, "Table 'employees.employes' doesn't exist")
[SQL: 
    select * from employes
    limit 5;
    ]
(Background on this error at: https://sqlalche.me/e/14/f405)

#### 6. Read the employees and titles tables into two separate DataFrames.



In [61]:
employees = pd.read_sql('''
    select * from employees
    ;
    ''', get_db_url('employees'))
employees

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [62]:
titles = pd.read_sql('''
    select * from titles
    ;
    ''', get_db_url('employees'))
titles

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

#### 7. How many rows and columns do you have in each DataFrame? Is that what you expected?

employees: 443308 rows x 4 cols, titles: 300024 rows x 6 cols. Multiple titles to different employees. Expected.

In [63]:
employees.shape

(300024, 6)

#### 8. Display the summary statistics for each DataFrame.


In [64]:
employees.describe()

emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

In [65]:
titles.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

#### 9. How many unique titles are in the titles DataFrame?



In [66]:
titles.title.unique()

array(['Senior Engineer', 'Staff', 'Engineer', 'Senior Staff',
       'Assistant Engineer', 'Technique Leader', 'Manager'], dtype=object)

In [67]:
titles.title.nunique()

7

#### 10. What is the oldest date in the to_date column?


In [68]:
titles.to_date.min()

datetime.date(1985, 3, 1)

In [69]:
titles.to_date.sort_values().head()

16064     1985-03-01
406358    1985-03-11
346216    1985-03-20
35527     1985-03-23
277603    1985-04-17
Name: to_date, dtype: object

#### 11. What is the most recent date in the to_date column?


In [70]:
titles.to_date.max()

datetime.date(9999, 1, 1)

In [71]:
titles.to_date[titles.to_date != titles.to_date.max()].max()

datetime.date(2002, 8, 1)

# Exercises II

#### 1. Copy the users and roles DataFrames from the examples above.


In [72]:
# Create the users DataFrame.

users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users


id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [73]:
# Create the roles DataFrame

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles


id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

#### 2. What is the result of using a right join on the DataFrames?



In [74]:
users.merge(roles, left_on='role_id', right_on='id', how='right', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0     1      admin        both
1   2.0    joe      2.0     2     author        both
2   3.0  sally      3.0     3   reviewer        both
3   4.0   adam      3.0     3   reviewer        both
4   NaN    NaN      NaN     4  commenter  right_only

#### 3. What is the result of using an outer join on the DataFrames?


In [75]:
users.merge(roles, left_on='role_id', right_on='id', how='outer', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0   1.0      admin        both
1   2.0    joe      2.0   2.0     author        both
2   3.0  sally      3.0   3.0   reviewer        both
3   4.0   adam      3.0   3.0   reviewer        both
4   5.0   jane      NaN   NaN        NaN   left_only
5   6.0   mike      NaN   NaN        NaN   left_only
6   NaN    NaN      NaN   4.0  commenter  right_only

#### 4. What happens if you drop the foreign keys from the DataFrames and try to merge them?





In [76]:
users.merge(roles, how='right', indicator=True)

id       name  role_id      _merge
0   1      admin      NaN  right_only
1   2     author      NaN  right_only
2   3   reviewer      NaN  right_only
3   4  commenter      NaN  right_only

#### 5. Load the mpg dataset from PyDataset.



In [77]:
mpg = data('mpg')
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact

In [78]:
type(mpg)

pandas.core.frame.DataFrame

#### 6. Output and read the documentation for the mpg dataset.



In [79]:
data('mpg', show_doc=True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




#### 7. How many rows and columns are in the dataset?
234 rows and 11 columns


In [80]:
mpg.shape

(234, 11)

#### 8. Check out your column names and perform any cleanup you may want on them.



In [81]:
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact

In [82]:
mpg.columns = ['manufacturer', 'model', 'displacement', 'year',
              'cylinders', 'transmission', 'drive_wheels',
               'city_mileage', 'hwy_mileage', 'fl', 'class']

#### 9. Display the summary statistics for the dataset.



In [83]:
mpg.describe()

displacement         year   cylinders  city_mileage  hwy_mileage
count    234.000000   234.000000  234.000000    234.000000   234.000000
mean       3.471795  2003.500000    5.888889     16.858974    23.440171
std        1.291959     4.509646    1.611534      4.255946     5.954643
min        1.600000  1999.000000    4.000000      9.000000    12.000000
25%        2.400000  1999.000000    4.000000     14.000000    18.000000
50%        3.300000  2003.500000    6.000000     17.000000    24.000000
75%        4.600000  2008.000000    8.000000     19.000000    27.000000
max        7.000000  2008.000000    8.000000     35.000000    44.000000

#### 10. How many different manufacturers are there?



In [84]:
len(mpg.manufacturer.value_counts())

15

In [85]:
mpg.manufacturer.unique()

array(['audi', 'chevrolet', 'dodge', 'ford', 'honda', 'hyundai', 'jeep',
       'land rover', 'lincoln', 'mercury', 'nissan', 'pontiac', 'subaru',
       'toyota', 'volkswagen'], dtype=object)

In [86]:
mpg.manufacturer.nunique()

15

#### 11. How many different models are there?



In [87]:
len(mpg.model.value_counts())

38

#### 12. Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.



In [88]:
mpg['mileage_difference'] = mpg.hwy_mileage - mpg.city_mileage
mpg

manufacturer   model  displacement  year  cylinders transmission  \
1           audi      a4           1.8  1999          4     auto(l5)   
2           audi      a4           1.8  1999          4   manual(m5)   
3           audi      a4           2.0  2008          4   manual(m6)   
4           audi      a4           2.0  2008          4     auto(av)   
5           audi      a4           2.8  1999          6     auto(l5)   
..           ...     ...           ...   ...        ...          ...   
230   volkswagen  passat           2.0  2008          4     auto(s6)   
231   volkswagen  passat           2.0  2008          4   manual(m6)   
232   volkswagen  passat           2.8  1999          6     auto(l5)   
233   volkswagen  passat           2.8  1999          6   manual(m5)   
234   volkswagen  passat           3.6  2008          6     auto(s6)   

    drive_wheels  city_mileage  hwy_mileage fl    class  mileage_difference  
1              f            18           29  p  compact                  11  
2              f            21           29  p  compact                   8  
3              f            20           31  p  compact                  11  
4              f            21           30  p  compact                   9  
5              f            16           26  p  compact                  10  
..           ...           ...          ... ..      ...                 ...  
230            f            19           28  p  midsize                   9  
231            f            21           29  p  midsize                   8  
232            f            16           26  p  midsize                  10  
233            f            18           26  p  midsize                   8  
234            f            17           26  p  midsize                   9  

[234 rows x 12 columns]

#### 13. Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.




In [89]:
mpg['average_mileage2'] = mpg[['hwy_mileage', 'city_mileage']].agg('mean',axis=1)
mpg

manufacturer   model  displacement  year  cylinders transmission  \
1           audi      a4           1.8  1999          4     auto(l5)   
2           audi      a4           1.8  1999          4   manual(m5)   
3           audi      a4           2.0  2008          4   manual(m6)   
4           audi      a4           2.0  2008          4     auto(av)   
5           audi      a4           2.8  1999          6     auto(l5)   
..           ...     ...           ...   ...        ...          ...   
230   volkswagen  passat           2.0  2008          4     auto(s6)   
231   volkswagen  passat           2.0  2008          4   manual(m6)   
232   volkswagen  passat           2.8  1999          6     auto(l5)   
233   volkswagen  passat           2.8  1999          6   manual(m5)   
234   volkswagen  passat           3.6  2008          6     auto(s6)   

    drive_wheels  city_mileage  hwy_mileage fl    class  mileage_difference  \
1              f            18           29  p  compact                  11   
2              f            21           29  p  compact                   8   
3              f            20           31  p  compact                  11   
4              f            21           30  p  compact                   9   
5              f            16           26  p  compact                  10   
..           ...           ...          ... ..      ...                 ...   
230            f            19           28  p  midsize                   9   
231            f            21           29  p  midsize                   8   
232            f            16           26  p  midsize                  10   
233            f            18           26  p  midsize                   8   
234            f            17           26  p  midsize                   9   

     average_mileage2  
1                23.5  
2                25.0  
3                25.5  
4                25.5  
5                21.0  
..                ...  
230              23.5  
231              25.0  
232              21.0  
233              22.0  
234              21.5  

[234 rows x 13 columns]

In [90]:
mpg['average_mileage'] = mpg[['hwy_mileage', 'city_mileage']].mean(axis=1)
mpg

manufacturer   model  displacement  year  cylinders transmission  \
1           audi      a4           1.8  1999          4     auto(l5)   
2           audi      a4           1.8  1999          4   manual(m5)   
3           audi      a4           2.0  2008          4   manual(m6)   
4           audi      a4           2.0  2008          4     auto(av)   
5           audi      a4           2.8  1999          6     auto(l5)   
..           ...     ...           ...   ...        ...          ...   
230   volkswagen  passat           2.0  2008          4     auto(s6)   
231   volkswagen  passat           2.0  2008          4   manual(m6)   
232   volkswagen  passat           2.8  1999          6     auto(l5)   
233   volkswagen  passat           2.8  1999          6   manual(m5)   
234   volkswagen  passat           3.6  2008          6     auto(s6)   

    drive_wheels  city_mileage  hwy_mileage fl    class  mileage_difference  \
1              f            18           29  p  compact                  11   
2              f            21           29  p  compact                   8   
3              f            20           31  p  compact                  11   
4              f            21           30  p  compact                   9   
5              f            16           26  p  compact                  10   
..           ...           ...          ... ..      ...                 ...   
230            f            19           28  p  midsize                   9   
231            f            21           29  p  midsize                   8   
232            f            16           26  p  midsize                  10   
233            f            18           26  p  midsize                   8   
234            f            17           26  p  midsize                   9   

     average_mileage2  average_mileage  
1                23.5             23.5  
2                25.0             25.0  
3                25.5             25.5  
4                25.5             25.5  
5                21.0             21.0  
..                ...              ...  
230              23.5             23.5  
231              25.0             25.0  
232              21.0             21.0  
233              22.0             22.0  
234              21.5             21.5  

[234 rows x 14 columns]

#### 14. Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission.




In [91]:
# mpg['is_automatic'] = mpg.transmission.apply(lambda x: 'auto' in x)

In [92]:
# mpg['is_automatic'] = mpg.transmission.str.contains('auto')

In [93]:
mpg['is_automatic'] = mpg.transmission.str.startswith('auto')
mpg

manufacturer   model  displacement  year  cylinders transmission  \
1           audi      a4           1.8  1999          4     auto(l5)   
2           audi      a4           1.8  1999          4   manual(m5)   
3           audi      a4           2.0  2008          4   manual(m6)   
4           audi      a4           2.0  2008          4     auto(av)   
5           audi      a4           2.8  1999          6     auto(l5)   
..           ...     ...           ...   ...        ...          ...   
230   volkswagen  passat           2.0  2008          4     auto(s6)   
231   volkswagen  passat           2.0  2008          4   manual(m6)   
232   volkswagen  passat           2.8  1999          6     auto(l5)   
233   volkswagen  passat           2.8  1999          6   manual(m5)   
234   volkswagen  passat           3.6  2008          6     auto(s6)   

    drive_wheels  city_mileage  hwy_mileage fl    class  mileage_difference  \
1              f            18           29  p  compact                  11   
2              f            21           29  p  compact                   8   
3              f            20           31  p  compact                  11   
4              f            21           30  p  compact                   9   
5              f            16           26  p  compact                  10   
..           ...           ...          ... ..      ...                 ...   
230            f            19           28  p  midsize                   9   
231            f            21           29  p  midsize                   8   
232            f            16           26  p  midsize                  10   
233            f            18           26  p  midsize                   8   
234            f            17           26  p  midsize                   9   

     average_mileage2  average_mileage  is_automatic  
1                23.5             23.5          True  
2                25.0             25.0         False  
3                25.5             25.5         False  
4                25.5             25.5          True  
5                21.0             21.0          True  
..                ...              ...           ...  
230              23.5             23.5          True  
231              25.0             25.0         False  
232              21.0             21.0          True  
233              22.0             22.0         False  
234              21.5             21.5          True  

[234 rows x 15 columns]

#### 15. Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average?



In [94]:
mpg.groupby('manufacturer').average_mileage.mean().nlargest(1)

manufacturer
honda    28.5
Name: average_mileage, dtype: float64

#### 16. Do automatic or manual cars have better miles per gallon?



In [95]:
mpg.groupby('is_automatic').average_mileage.mean()

is_automatic
False    22.227273
True     19.130573
Name: average_mileage, dtype: float64

# Exercises III

#### 1. Use your get_db_url function to help you explore the data from the chipotle database.



In [96]:
url = get_db_url('chipotle')

In [97]:
pd.read_sql('''
    show tables
    ;
    ''', url)

Tables_in_chipotle
0             orders

In [98]:
pd.read_sql('''
    select * from orders
    limit 5;
    ''', url)

id  order_id  quantity                              item_name  \
0   1         1         1           Chips and Fresh Tomato Salsa   
1   2         1         1                                   Izze   
2   3         1         1                       Nantucket Nectar   
3   4         1         1  Chips and Tomatillo-Green Chili Salsa   
4   5         2         2                           Chicken Bowl   

                                                           choice_description  \
0                                                                         nan   
1                                                                [Clementine]   
2                                                                     [Apple]   
3                                                                         nan   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   

  item_price  
0     $2.39   
1     $3.39   
2     $3.39   
3     $2.39   
4    $16.98

In [99]:
df = pd.read_sql('''
    select * from orders
    ;''', url)
df

id  order_id  quantity                              item_name  \
0        1         1         1           Chips and Fresh Tomato Salsa   
1        2         1         1                                   Izze   
2        3         1         1                       Nantucket Nectar   
3        4         1         1  Chips and Tomatillo-Green Chili Salsa   
4        5         2         2                           Chicken Bowl   
...    ...       ...       ...                                    ...   
4617  4618      1833         1                          Steak Burrito   
4618  4619      1833         1                          Steak Burrito   
4619  4620      1834         1                     Chicken Salad Bowl   
4620  4621      1834         1                     Chicken Salad Bowl   
4621  4622      1834         1                     Chicken Salad Bowl   

                                                                     choice_description  \
0                                                                                   nan   
1                                                                          [Clementine]   
2                                                                               [Apple]   
3                                                                                   nan   
4            [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
...                                                                                 ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
4618               [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese, Lettuce, Guacamole]]   
4619         [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]   
4620                                 [Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]   
4621                    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]   

     item_price  
0        $2.39   
1        $3.39   
2        $3.39   
3        $2.39   
4       $16.98   
...         ...  
4617    $11.75   
4618    $11.75   
4619    $11.25   
4620     $8.75   
4621     $8.75   

[4622 rows x 6 columns]

In [100]:
df[df.item_name == 'Chicken Bowl']

id  order_id  quantity     item_name  \
4        5         2         2  Chicken Bowl   
5        6         3         1  Chicken Bowl   
13      14         7         1  Chicken Bowl   
19      20        10         1  Chicken Bowl   
26      27        13         1  Chicken Bowl   
...    ...       ...       ...           ...   
4590  4591      1825         1  Chicken Bowl   
4591  4592      1825         1  Chicken Bowl   
4595  4596      1826         1  Chicken Bowl   
4599  4600      1827         1  Chicken Bowl   
4604  4605      1828         1  Chicken Bowl   

                                                                             choice_description  \
4                    [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
5                   [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sour Cream, Guacamole, Lettuce]]   
13               [Fresh Tomato Salsa, [Fajita Vegetables, Rice, Cheese, Sour Cream, Guacamole]]   
19    [Tomatillo Red Chili Salsa, [Fajita Vegetables, Black Beans, Sour Cream, Cheese, Lettu...   
26    [Roasted Chili Corn Salsa (Medium), [Pinto Beans, Rice, Fajita Veggies, Cheese, Sour C...   
...                                                                                         ...   
4590                [Roasted Chili Corn Salsa, [Rice, Black Beans, Cheese, Lettuce, Guacamole]]   
4591              [Tomatillo Red Chili Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce]]   
4595                                         [Tomatillo Green Chili Salsa, [Rice, Black Beans]]   
4599                                              [Roasted Chili Corn Salsa, [Cheese, Lettuce]]   
4604                              [Fresh Tomato Salsa, [Rice, Black Beans, Cheese, Sour Cream]]   

     item_price  
4       $16.98   
5       $10.98   
13      $11.25   
19       $8.75   
26       $8.49   
...         ...  
4590    $11.25   
4591     $8.75   
4595     $8.75   
4599     $8.75   
4604     $8.75   

[726 rows x 6 columns]

#### 2. What is the total price for each order?



In [118]:
df.dtypes

id                      int64
order_id                int64
quantity                int64
item_name              object
choice_description     object
item_price            float64
dtype: object

In [101]:
df.item_price = df.item_price.str.strip('$').astype(float)

In [102]:
df.quantity = df.quantity.astype(int)

In [119]:
df.groupby('order_id').sum().item_price
# same as following

order_id
1       11.56
2       16.98
3       12.67
4       21.00
5       13.70
        ...  
1830    23.00
1831    12.90
1832    13.20
1833    23.50
1834    28.75
Name: item_price, Length: 1834, dtype: float64

In [104]:
df.groupby('order_id').item_price.sum()
# same as previous

order_id
1       11.56
2       16.98
3       12.67
4       21.00
5       13.70
        ...  
1830    23.00
1831    12.90
1832    13.20
1833    23.50
1834    28.75
Name: item_price, Length: 1834, dtype: float64

#### 3. What are the most popular 3 items?



In [ ]:
### get bracket version of this from exercise review

In [120]:
df.groupby('item_name').quantity.sum().sort_values(ascending=False).head(3)
# same as following

item_name
Chicken Bowl           761
Chicken Burrito        591
Chips and Guacamole    506
Name: quantity, dtype: int64

In [105]:
df.groupby('item_name').quantity.sum().nlargest(3)
# same as previous

item_name
Chicken Bowl           761
Chicken Burrito        591
Chips and Guacamole    506
Name: quantity, dtype: int64

#### 4. Which item has produced the most revenue?



In [ ]:
# it looks like item_price accounts for quantity

In [106]:
df.pivot_table(index='item_name', values='item_price', aggfunc='sum').nlargest(5, columns='item_price')

item_price
item_name                      
Chicken Bowl            7342.73
Chicken Burrito         5575.82
Steak Burrito           3851.43
Steak Bowl              2260.19
Chips and Guacamole     2201.04

In [107]:
df.groupby('item_name').item_price.agg('sum').nlargest(1)
# same as following

item_name
Chicken Bowl    7342.73
Name: item_price, dtype: float64

In [121]:
df.groupby('item_name').item_price.sum().nlargest(1)
# same as previous

item_name
Chicken Bowl    7342.73
Name: item_price, dtype: float64

#### 5. Join the employees and titles DataFrames together.



In [111]:
emp_tit = employees.merge(titles, how='inner', on='emp_no')
emp_tit.head()

emp_no  birth_date first_name last_name gender   hire_date  \
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26   
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21   
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28   
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
4   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   

             title   from_date     to_date  
0  Senior Engineer  1986-06-26  9999-01-01  
1            Staff  1996-08-03  9999-01-01  
2  Senior Engineer  1995-12-03  9999-01-01  
3         Engineer  1986-12-01  1995-12-01  
4  Senior Engineer  1995-12-01  9999-01-01

#### 6. For each title, find the hire date of the employee that was hired most recently with that title.



In [123]:
emp_tit.groupby('title').hire_date.max()

title
Assistant Engineer    1999-12-24
Engineer              2000-01-28
Manager               1992-02-05
Senior Engineer       2000-01-01
Senior Staff          2000-01-13
Staff                 2000-01-12
Technique Leader      1999-12-31
Name: hire_date, dtype: object

In [122]:
emp_tit.groupby('title').hire_date.agg('max')

title
Assistant Engineer    1999-12-24
Engineer              2000-01-28
Manager               1992-02-05
Senior Engineer       2000-01-01
Senior Staff          2000-01-13
Staff                 2000-01-12
Technique Leader      1999-12-31
Name: hire_date, dtype: object

#### 7. Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)



In [114]:
url = get_db_url('employees')
dep_tit = pd.read_sql('''
    select t.title, d.dept_name
from titles t
	join dept_emp de
		using (emp_no)
	join departments d
		using (dept_no)
    ;''', url)

In [115]:
dep_tit

title         dept_name
0              Staff  Customer Service
1       Senior Staff  Customer Service
2              Staff  Customer Service
3       Senior Staff  Customer Service
4              Staff  Customer Service
...              ...               ...
489898  Senior Staff             Sales
489899         Staff             Sales
489900         Staff             Sales
489901  Senior Staff             Sales
489902         Staff             Sales

[489903 rows x 2 columns]

In [117]:
pd.crosstab(dep_tit.title, dep_tit.dept_name).T

title               Assistant Engineer  Engineer  Manager  Senior Engineer  \
dept_name                                                                    
Customer Service                   298      2362        4             2027   
Development                       7769     58135        2            49326   
Finance                              0         0        2                0   
Human Resources                      0         0        2                0   
Marketing                            0         0        2                0   
Production                        6445     49649        4            42205   
Quality Management                1831     13852        4            11864   
Research                           378      2986        2             2570   
Sales                                0         0        2                0   

title               Senior Staff  Staff  Technique Leader  
dept_name                                                  
Customer Service           13925  16150               309  
Development                 1247   1424              7683  
Finance                    12139  13929                 0  
Human Resources            12274  14342                 0  
Marketing                  13940  16196                 0  
Production                  1270   1478              6557  
Quality Management             0      0              1795  
Research                   11637  13495               393  
Sales                      36191  41808                 0